### ISSUE TITLE
Linked Account: GitLab

### ISSUE THREAD

**Issue by EricWittmann** [Apicurio/apicurio-studio#161]
Now that Apicurio supports account linking, add GitLab as one of the supported account types.  Currently GitLab is already included in the UI but disabled (with messages about support coming soon).

There are various changes required to add GitLab support:

* _Keycloak_: Configure a GitLab Identity Provider in Keycloak with alias "gitlab" (may require the latest version of KC)
* _UI_: Update the **Create API** page to enable the GitLab button.  Clicking that button should display GitLab-specific form fields similar to how the current GitHub form fields work.
* _UI_: Update the **Settings>Accounts** page to enable GitLab account linking
* _Back-End_: Add a GitLab implementation of a [ISourceConnector](https://github.com/Apicurio/apicurio-studio/blob/master/back-end/hub-api/src/main/java/io/apicurio/hub/api/connectors/ISourceConnector.java)
* _Back-End_: Add the GitLab source connector to the [SourceConnectorFactory](https://github.com/Apicurio/apicurio-studio/blob/master/back-end/hub-api/src/main/java/io/apicurio/hub/api/connectors/SourceConnectorFactory.java)

**Comment by EricWittmann** [comment]
Note that the source connector is the class that actually connects Apicurio to GitLab.  It should basically consist of a bunch of REST calls to the [GitLab API](https://docs.gitlab.com/ee/api/README.html) and will only function if the user has a linked GitLab account.

**Comment by jonasrob** [comment]
@EricWittmann I am collaborating on this story and have a quick question for you.  Settings > Accounts  is all set.  For Create API we are at the point where Group (Group is to GitLab as Organization is to GitHub) and Repository drop downs are populated.  Wonder if we should provide a drop down for the  branch as well?  Perhaps even go as far as populating this dropdown with the default branch.  That way we commit to a specific branch and don't assume the default branch.  Thoughts?

**Comment by EricWittmann** [comment]
Yeah populating the branch makes a lot of sense.  I should do that for GitHub then too.

Note taht ultimately I want to break apart the link between the API and where it is stored, but that's further down the roadmap.  What I mean is that at some point I'm planning on storing the API content in Apicurio and having an **action** to publish it in a source code repository.  Not relevant necessarily to what you're doing in this task, but FYI anyway. :)

**Comment by mccrackend** [comment]
Sorry to hijack this issue, but this really interested me: 

> at some point I'm planning on storing the API content in Apicurio and having an action to publish it in a source code repository.

This is really cool - I could really see myself **wanting** to use an app that gives me:
1. a sandbox editor for API designs
1. the ability to pull in existing designs from remote repos
1. the ability to commit designs back to specific locations in remote repos

But would I ever not want my sandbox-type scribblings to make it into a repo's branch where I'm experimenting?  Maybe in the case where I don't have a repo created for a project yet.

I'll move future discussion of this over to issue #85 that I just discovered  :)

**Comment by jonasrob** [comment]
@EricWittmann Dan Mikita mentioned that you might be interested in picking up the gitLab portion where I have left off. My apologies on the delay.  He said that you would like the gitLab integration in the next release.  

I have invited you to my local branch where we are at.  We are working through some local issues at the moment and that has delayed progress.

https://github.com/jonasrob/apicurio-studio/commit/da4408d27f4e50f98db087f0604c9dfd04048fcd

I rolled all of our work into the commit above.   Not sure when you are planning on releasing, but I am pretty close to addressing my local environment issues, and resuming work.

Please let me know how you would like to proceed.

**Comment by EricWittmann** [comment]
OK great thanks!  I'm at a conference this week, so I probably won't have a ton of time to look at stuff until I get back.  I don't have a specific deadline for release - I just want to tie some things off asap so I can move on to Beta 2 (which will feature multi-user concurrent editing). :)

**Comment by jonasrob** [comment]
Hi @EricWittmann my local environment is up.  It isn't great, but I am going to give finishing the GitLab feature a shot.  I cleaned up the client side code last night, and am going to attack the remaining api over today and tomorrow.  Figured at the very least I can update the jonasrob branch by EOD tomorrow, or at the most have a PR ready for ya.  Please let me know if you have resumed work in parallel.

**Comment by jonasrob** [comment]
@EricWittmann I didn't quite make it all the way through.  Please see https://github.com/jonasrob/apicurio-studio/tree/gitLab for updates.  Made a lot of progress this weekend, the commit call to GitLab isn't working via apicurio-studio, but same call works in Postman.  Let me know where you would like to go from here.  Wouldn't mind continuing the work, don't want to slow you down either.

**Comment by EricWittmann** [comment]
Thanks @jonasrob !  I'll try to have a look at this stuff today and let you know.  I'm hoping that I can get close to a release this week.  But another option may be for me to keep Beta 1 on **master** and simply branch for Beta 2.  That might be the smarter move in any case.  In that case we can all continue to push Beta 1 along incrementally. :)

I'll report back here later.

**Comment by EricWittmann** [comment]
OK I've gone through the diff and a lot of this looks really good.  I don't have any idea yet why the commit call to GitLab isn't working for you (I haven't reproduced that yet).  I'm going to set up my local environment so I can do some testing, and then create a junit test for the **GitLabSourceConnector**.  That should make it easier to figure out what's going on.  :)  I'll report back later today.

**Comment by EricWittmann** [comment]
Hey @jonasrob - I have some news to report.  :)

Assuming you're hitting the same obstacle that I'm hitting, the root cause can be traced to the HTTP client we're using (mashape's Unirest client).  That HTTP client is doing some funky URL manipulation before it creates the underlying Apache HTTP Client object.  Here is a reference if interested:

https://github.com/Mashape/unirest-java/blob/master/src/main/java/com/mashape/unirest/http/HttpClientHelper.java#L171-L181

As you can see, line 172 is the culprit.  It's doing a **URLDecoder.decode()** on the path.  This converts the GitLab URL-encoded values (for both the Group/Project name as well as the resource path) back from being URL encoded to **NOT** being URL encoded.  :(

Basically it mucks up the URL into something that GitLab no longer understands.

I'm going to try and double-encode those aspects of the URL to see if that works.  Otherwise we might need to use the Apache HTTP Client directly, instead of Unirest...

**Comment by EricWittmann** [comment]
Update: I don't think there is a way to work around this issue in the Unirest client, so I think the only option is to switch to using Apache Http Client directly.  Bummer.  I'll work on that tomorrow and send you a patch for your branch.

**Comment by jonasrob** [comment]
@EricWittmann thoughts on keeping Unirest and looking up the projects id instead?  Would be an extra call to the GitLab API, but wouldn't be a big deal.

**Comment by EricWittmann** [comment]
Yeah that would help for the project ID but the resource path still needs to be encoded.  Don't think there's a way around that, is there?

**Comment by jonasrob** [comment]
Gotcha, yip sure does.. bummer :)  Thanks for looking into this further.

**Comment by EricWittmann** [comment]
Alright.  I've swapped out Unirest for straight-up Apache HTTP client.  Using a junit test (@Ignored by default) I've confirmed that most of the methods in **GitLabSourceConnector** now work as expected.  I have a patch for you to apply to your local **gitLab** branch.  It should be attached to this comment.  Apply the patch via:

```
$ cd ~/git/apicurio/apicurio-studio
$ git apply gitlab.patch.txt
```

Let me know if that works out OK.  The patch is larger than it needs to be, so I apologize for the noise.  I added some missing copyright notices and removed some unused imports to clear up Eclipse warnings. 

[gitlab.patch.txt](https://github.com/Apicurio/apicurio-studio/files/1353277/gitlab.patch.txt)

I have **NOT** tested the UI at all.  I looked at the changes you made (via GitHub's diff tools) and they seem good.

Let me know what you want to do as the next step.  If you want to apply the patch, test, and submit a PR that's great.  If there is more you want/need to do first that's fine too, just let me know. :)

**Comment by jonasrob** [comment]
@EricWittmann another crazy week.  I will be taking a look over the weekend.  Thanks for the patch, much appreciated.

**Comment by EricWittmann** [comment]
OK sounds good.  Happy Friday and have a nice weekend. :)

**Comment by EricWittmann** [comment]
@jonasrob Hey there.  FYI I have started work on getting the gitlab support finalized and committed.  I think everything is working (using a local version of Keycloak's latest release candidate) locally.  Now I just need to rebase it on master and fix the inevitable conflicts.  Just letting you know I've started on this so we don't duplicate anything!
